### PR TITLE
 Eliminate redundant sorts in BWD pass by caching indexes between embedding chunks 

### PR DESCRIPTION
**Pull request by Shuwen-Fang**
# Description

Optimizes token routing in `ring_ragged_sort` during embedding chunking by eliminating redundant `jnp.argsort` calls in both the forward and backward passes.

### Performance Impact
* **Latency Reduction**: Saves **~600 µs per chunk per layer** in the backward pass (see [Profile Breakdown Screenshot](https://screenshot.googleplex.com/7DHqVQXFy2P5Kb3)).

### Regression test 
ran [script](https://paste.googleplex.com/5017217979777024) with deepseek tiny on VM on main branch and PR's branch with no significant tokens/s/device difference: [main branch logs](https://paste.googleplex.com/5000188837101568), [PR branch logs](https://paste.googleplex.com/5017590291365888)

### How & Why Sort Is Removed
All embedding chunks within a layer share identical token routing (topk_indices_2d and valid_rows_mask), but every embedding chunk repeatedly recomputed the SparseCore valid-row index permutation via jnp.argsort in _preprocess. This PR caches the sorted by validity array by computing it once during the forward pass, then passing it into future bwd passes via custom VJP residuals _ring_ragged_sort_bwd, so that jnp.argsort runs once per layer across all forward and backward chunk ops.

---

# Tests

1. **Unit Test Verification**:
   * Executed `test_moe_emb_chunking_random_routing` in `tests/unit/moe_test.py` on TPU (`tpu7x-128`) on VM:
   ```text
   tests/unit/moe_test.py::RoutedMoeTest::test_moe_emb_chunking_random_routing . [100%]
   ================ 1 passed, 41 deselected in 40.00s ================
   ```

2. **Loss & Perplexity Correctness Run (`num_moe_emb_chunks=4`)**:
   * Executed 20-step pretraining correctness runs (`tmp/run_correctness.sh`) comparing `main` branch vs `sort_index_once` branch with 4 embedding chunks active:
     * **Step 0 Loss**: `12.258` (`main`) vs `12.258` (`sort_index_once`)
     * **Step 10 Loss**: `11.985` (`main`) vs `11.985` (`sort_index_once`)
     * **Step 19 Loss**: `11.919` (`main`) vs `11.919` (`sort_index_once`)
     * **Step 19 Perplexity**: `150053.750` (`main`) vs `150053.750` (`sort_index_once`)
   * Confirmed **100% numerical equivalence** across all 20 steps under chunked execution.

3. **Sort Elimination Verification via Profile**:
   * Deployed XManager experiment [XID 273744286](http://xids/273744286) (`deepseekv3-fsdp512-ep8-chunked-2`) to verify via profile trace that `argsort` overhead in the backward pass was eliminated.

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests and provided workload links above if applicable.

### CODE REVIEW

File: src/maxtext/kernels/ragged/ragged_gather_reduce_v2.py [hunk 1]
```diff
@@ -279,30 +302,20 @@ def _preprocess(
...
+    sorted_by_validity = _compute_sorted_by_validity(valid_rows_mask_2d, num_row_partitions, row_chunk_size)
 
   num_src_rows_per_row_partition = jnp.pad(
       jnp.sum(valid_rows_mask_2d, axis=-1).astype(jnp.int32),
       (0, max(0, num_simd_lanes - num_row_partitions)),
   )
   mask = jnp.any(valid_rows_mask.reshape(-1, reduce_group_size), axis=-1)
   return (
-      sorted_by_validity.astype(jnp.int32),
+      sorted_by_validity,
```
Comment (NuojCheng) [inline src/maxtext/kernels/ragged/ragged_gather_reduce_v2.py]: we need jnp.int32 for accumulation, otherwise got regression issue

File: src/maxtext/layers/moe.py [hunk 2]
```diff
@@ -846,11 +847,12 @@ def permute(
       self,
       inputs,
       gate_logits,
-      pre_bias_logits,
+      pre_bias_logits=None,
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py]: emmm not clear why make this change

**Review by NuojCheng — COMMENTED** [review]
1. You could add a test protecting your feature.
2. Could you add a regression test using flags in [here](https://source.corp.google.com/piper///depot/google3/experimental/users/chengnuojin/maxtext-xm/deepseek3-671b/2026-06-mlperf/ds-fsdp64-ep8-splash.sh)? I would like to make sure our performance is not impacted by this change. Feel free to reduce layers and chip sizes for easier repro.
